File: 2024/src/day03.fs
module Day03

open System.Text.RegularExpressions

let runPartOne input =
    Regex.Matches(input, "mul\((\d+),(\d+)\)")
    |> Seq.sumBy (fun m -> int m.Groups[1].Value * int m.Groups[2].Value)

let runPartTwo input =
    ((0, true), Regex.Matches(input, "(?<do>do\(\))|(?<dont>don't\(\))|mul\((?<a>\d+),(?<b>\d+)\)"))
    ||> Seq.fold (fun (sum, todo) m ->
        match (m.Groups["do"].Success, m.Groups["dont"].Success, todo) with
        | true, _, _ -> (sum, true)
        | _, true, _ -> (sum, false)
        | _, _, true -> (sum + int m.Groups["a"].Value * int m.Groups["b"].Value, true)
        | _ -> (sum, todo))
    |> fst

File: 2024/src/day03.ts
export function runPartOne(input: string) {
  return [...input.matchAll(/mul\((\d+),(\d+)\)/g)].reduce(
    (sum, m) => sum + parseInt(m[1]) * parseInt(m[2]),
    0,
  );
}

export function runPartTwo(input: string): number {
  return Array.from(
    input.matchAll(
      /(?<do>do\(\))|(?<dont>don't\(\))|mul\((?<a>\d+),(?<b>\d+)\)/g,
    ),
  ).reduce<[number, boolean]>(
    ([sum, todo], m) => {
      if (m.groups!.do) return [sum, true];
      else if (m.groups!.dont) return [sum, false];
      else if (todo)
        return [sum + parseInt(m.groups!.a) * parseInt(m.groups!.b), todo];
      return [sum, todo];
    },
    [0, true],
  )[0];
}
